File: app/src/main/java/biz/dealnote/xmpp/db/interfaces/IContactsRepository.java
package biz.dealnote.xmpp.db.interfaces;

import android.support.annotation.NonNull;

import org.jivesoftware.smackx.vcardtemp.packet.VCard;

import biz.dealnote.xmpp.model.Contact;
import biz.dealnote.xmpp.util.Optional;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;

/**
 * Created by ruslan.kolbasa on 02.11.2016.
 * phoenix_for_xmpp
 */
public interface IContactsRepository {

    Single<Optional<Contact>> findById(int id);

    Single<Optional<Contact>> findByJid(@NonNull String jid);

    Completable upsert(String bareJid, VCard vCard);

    Single<Integer> getContactIdPutIfNotExist(String bareJid);

    Single<Contact> getByJid(String bareJid);

    Flowable<Contact> observeAdding();

    Flowable<Contact> observeUpdates();

    byte[] findPhotoByHash(String hash);
}

File: app/src/main/java/biz/dealnote/xmpp/repo/IContactsRepository.kt
package biz.dealnote.xmpp.repo

import biz.dealnote.xmpp.model.Contact
import biz.dealnote.xmpp.model.User
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import org.jivesoftware.smack.roster.RosterEntry
import org.jxmpp.jid.Jid

interface IContactsRepository {
    fun actualizeUserAndGet(account: Int, jid: String): Single<User>

    fun getContacts(): Single<List<Contact>>

    fun observeAddings(): Flowable<List<String>>

    fun observeVcards(): Flowable<Collection<User>>

    fun observeDeleting(): Flowable<List<String>>

    fun handleContactsAdded(account: Int, contacts: Collection<RosterEntry>): Completable

    fun handleContactsDeleted(account: Int, jids: Collection<Jid>): Completable
}